core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3242,6 +3209,133 @@ class KafkaApis(val requestChannel: RequestChannel,
+      // is to disambiguate structural errors in the envelope request itself, such
+      // as an invalid client address.
+
+      val forwardedRequestBuffer = envelope.requestData.duplicate()
+      val forwardedRequestHeader = parseForwardedRequestHeader(forwardedRequestBuffer).getOrElse {
+        throw new UnsupportedVersionException("Failed to parse request header from envelope")
+      }
+
+      val forwardedApi = forwardedRequestHeader.apiKey
+      if (!forwardedApi.forwardable || !forwardedApi.isEnabled) {
+        throw new InvalidRequestException(s"API $forwardedApi is not enabled or is not eligible for forwarding")
+      }
+
+      val forwardedContext = new RequestContext(
+        forwardedRequestHeader,
+        request.context.connectionId,
+        forwardedClientAddress,
+        forwardedPrincipal,
+        request.context.listenerName,
+        request.context.securityProtocol,
+        ClientInformation.EMPTY,
+        request.context.fromPrivilegedListener
+      )
+
+      val forwardedRequest = parseForwardedRequest(request, forwardedContext, forwardedRequestBuffer).getOrElse {
+        throw new UnsupportedVersionException(s"Failed to parse forwarded request with header $forwardedRequestHeader")
+      }
+
+      handle(forwardedRequest)
+    } catch {
+      case e: KafkaException =>
+        debug(s"Failed to handle envelope request $request", e)
+        sendErrorResponseMaybeThrottle(request, e)
+    }
+  }
+
+  private def parseForwardedClientAddress(
+    address: Array[Byte]
+  ): Option[InetAddress] = {
+    try {
+      Some(InetAddress.getByAddress(address))
+    } catch {
+      case e: UnknownHostException =>
+        None
+    }
+  }
+
+  private def parseForwardedRequest(
+    envelope: RequestChannel.Request,
+    forwardedContext: RequestContext,
+    buffer: ByteBuffer
+  ): Option[RequestChannel.Request] = {
+    try {
+      Some(new RequestChannel.Request(
+        processor = envelope.processor,
+        context = forwardedContext,
+        startTimeNanos = envelope.startTimeNanos,
+        envelope.memoryPool,
+        buffer,
+        requestChannel.metrics,
+        Some(envelope)
+      ))
+    } catch {
+      case e: InvalidRequestException =>
+        None
+    }
+  }
+
+  private def parseForwardedRequestHeader(
+    buffer: ByteBuffer
+  ): Option[RequestHeader] = {
+    try {
+      Some(RequestHeader.parse(buffer))
+    } catch {
+      case e: InvalidRequestException =>
+        None
+    }
+  }
+
+  private def parseForwardedPrincipal(
+    envelopeRequest: RequestChannel.Request
+  ): Option[KafkaPrincipal] = {
+    val envelope = envelopeRequest.body[EnvelopeRequest]
+    try {
+      val principalSerde = envelopeRequest.context.principalSerde.asScala
+      principalSerde.map { serde =>
+        serde.deserialize(envelope.requestPrincipal)
+      }
+    } catch {
+      case e: Exception =>
+        warn(s"Failed to deserialize principal from envelope request $envelope", e)
```

Comment:
We can covert the exception to ```throw new PrincipalDeserializationException("Failed to parse client principal from envelope")``` here to avoid redundant try-catch (another is in handleEnvelope) and unnecessary ```Option``` wrap.

Reply:
Thanks, makes sense and simplifies the code a bit.